### ISSUE TITLE
Implement Epoch Sync

### ISSUE THREAD

**Issue by SkidanovAlex** [near/nearcore#3488]
This issue tracks adding support for light speed sync.

**Comment by Kouprin** [comment]
@bowenwang1996 there is discussion I've had with @SkidanovAlex and want to share and synchronize with you our next steps.

-- TL;DR starts here --

This Tuesday I suggested to have Epoch Sync at the start + passive GC headers concept. It resolves two main tasks we want to achieve:
1. making syncing procedure fast;
2. clearing headers to control amount of memory / storage required for them.

To obtain that, I implemented the invariant: "store the headers from Epoch Sync up to the Head". However, Alex proposes different one: "store the headers from the Tail up to the Head". The motivation here is to reach maximum reliability on headers absence. When we have headers starting from Epoch Sync, we have most of the necessary headers all the time - and what's why it will be extremely hard to catch and reproduce any issues related to headers absence. Another thing about it, the invariant I've suggested is too broad and allows different nodes to use very different amount of storage / memory under the same invariant.

Alex's suggestion, basically, means enabling GC headers in realtime + has several important implications.
1. It doesn't lead us to Syncing state machine simplification that was proposed originally with Epoch Sync design. Instead, it needs to update dangerous method reset_data_pre_state_sync (and maybe some more) and keep having deal with it.
2. As far we can't go to the Syncing state machine simplification, there is no actual reason to run Epoch Sync at node start only. Instead, we can run Epoch Sync every time when we need to run State Sync. This is very convenient for our validators because they will not be required to restart their nodes to run Epoch Sync.

In terms of business decisions, I'm okay with both solutions. Unfortunately, there is no way to combine them in improving manner - we have to decide which approach we should use.

-- TL;DR ends here --

Personally, I trust Alex's vision and experience, that's why I prefer to follow his idea. However, it also means that I need to return to redesign / implementation status in my current work. In this case all approximate dates that I brought to you will be shifted.

**Comment by bowenwang1996** [comment]
@Kouprin @SkidanovAlex I would really love to get rid of `reset_data_pre_state_sync`, but I also agree with the invariant that Alex proposed. Could we reset tail to epoch sync start and delete everything in between instead?

**Comment by Kouprin** [comment]
@bowenwang1996 If you want to simply get rid of `reset_data_pre_state_sync`, please explain how you see the new policy of switching to the State Sync. Originally, we planned to disallow to run State Sync in the state machine more than once.

For now it seems to me that you suggest to enable GC of headers (that shifts the timeline anyway because we need to support it, fix the tests, migrate properly, etc.) + disallow to run Epoch Sync more than once. It takes bad from both approaches.

BTW, Epoch Sync is already clearing all data in the beginning and resets Tails.

**Comment by bowenwang1996** [comment]
@Kouprin I misunderstood the problem and have updated my comment

**Comment by bowenwang1996** [comment]
@Kouprin is it correct to say that with @SkidanovAlex's proposal, the state machine looks like:
```
epoch sync --> header sync --> state sync --> block sync
```
?

**Comment by Kouprin** [comment]
`epoch sync -> state sync -> header sync -> block sync`

**Comment by bowenwang1996** [comment]
@Kouprin I see, but it seems to me that we can do state sync and header sync in parallel?

**Comment by Kouprin** [comment]
@bowenwang1996 I'm trying to reuse `ColBlockPerHeight` as you suggested and it doesn't look too easy.

Basically, the meaning of `ColBlockPerHeight` is to have hashes of all Blocks that are available in the Store. The logic of `ColHeaderHashesByHeight` is to have hashes of all Headers that are available in the Store. The main difference appears in GC: `ColBlockPerHeight` is updated every time when the Block is deleted, but `ColHeaderHashesByHeight` is updated only while Chunks are deleting. 

Trying to merge both into one leads me to removing data from `ColBlockPerHeight` as late as `ColHeaderHashesByHeight` should be removed. It means, we cannot rely on and reuse `get_all_block_hashes_by_height()` to understand which Blocks are available. To avoid this problem, I need to Store actual Blocks mapping somewhere -> and here we reach to the initial issue.

I guess it's still possible to pack both into one. However, I'm not sure this is reasonable solution.

**Comment by Kouprin** [comment]
> @Kouprin I see, but it seems to me that we can do state sync and header sync in parallel?

Sorry for late response. Yes, we even can do Block Sync in parallel. We need to make sure the optimizations will help us with bottlenecks first.

**Comment by bowenwang1996** [comment]
> but ColHeaderHashesByHeight is updated only while Chunks are deleting.

Why is that?

**Comment by Kouprin** [comment]
> Why is that?

It's how GC Headers is designed. As we currently rely that all Headers are available, the Header shouldn't be deleted until any other data exist that may refer to it.

**Comment by bowenwang1996** [comment]
@Kouprin I see, but why do we need to gc `ColHeaderHashesByHeight` only when deleting chunk data. I looked at `clear_chunk_data` and it doesn't rely on anything related to the block that it is included in.

**Comment by SkidanovAlex** [comment]
@bowenwang1996 the way I understand the separation of the two columns is that `HeaderHashesByHeight` must be GCed the very last (because it is used to actually fetch the headers to be GCed, and the headers naturally must be garbage collected after everything else at that height was GCed).

`BlockPerHeight` today has an invariant that if a hash is present there, the block also exists.  
So the only way I see to unify the two is to make all the users of `BlockPerHeight` not to rely on the block existing. It is without a doubt doable, but is error prone (e.g. from a glance at `clear_forks_data`, it looks like it would just silently fail if a hash of a block was in `BlockPerHeight`, but not in `Blocks`).

I am in favor of not unifying them yet.

**Comment by Kouprin** [comment]
It's still possible to update `get_all_block_hashes_by_height` by calling `get_block` for each record and keep only ones that return no error. Not sure how easy it will work with several mutable self calls in Rust.

My biggest concern about unification is it seems we're adding complex and unclear essences with no particular reason. Saving little memory and several lines of code, we may introduce the logic no one wants to deal with in future. Having several such things, it will be frustrating and very hard to get into details and think about them properly.

**Comment by bowenwang1996** [comment]
Okay let's not unify the two columns then

**Comment by nikurt** [comment]
cc: @nikurt
